modules/services/wlsunset.nix

```diff
@@ -91,7 +124,11 @@ in {
             "-L ${cfg.longitude}"
             "-t ${toString cfg.temperature.night}"
             "-T ${toString cfg.temperature.day}"
+            "-S ${cfg.time.sunrise}"
+            "-s ${cfg.time.sunset}"
```

Comment:
Isn't this redundant with `latitute`/`longitude`?

I understand wanting to use fixed times rather than location-based schedules, but then I think all the options should be `type = nullOr str` and only given to `wlsunset` _if_ they're defined.

Reply:
I like that idea, it would be way better to do that instead actually.
Something like this: 
```nix
sunset = mkOption {
  type = with types; nullOr str;
  description = ''
     The time when the sun sets (in 24 hour format).
  '';
};
```

Reply:
Sorry for the late reply. Yeah I'm trying to work out how to do that without making it very complicated, I've been working on doing it like this.
Although I am stuck in terms of making sure either `latitude and longitude` or `sunrise and sunset` are present, as I believe that that will add a lot of extra code to the module.
```nix
{
  systemd.user.services.wlsunset.Service = {
     ExecStart = let
        args = [
           "-t ${toString cfg.temperature.night}"
           "-T ${toString cfg.temperature.day}"
           "-g ${cfg.gamma}"
           ((mkIf cfg.latitude != null) "-l ${cfg.latitude}")
           ((mkIf cfg.longitude != null) "-L ${cfg.longitude}")
           ((mkIf cfg.time.sunrise != null) "-S ${cfg.time.sunrise}")
           ((mkIf cfg.time.sunset != null) "-s ${cfg.time.sunset}")
           ((mkIf cfg.time.duration != null) "-d ${toString cfg.time.duration}")
           ((mkIf cfg.output != null) "-o ${cfg.output}")
        ];
       in "${cfg.package}/bin/wlsunset ${concatStringsSep " " args}";
   };
}
```

Reply:
I think you may be able to use `cli.toGNUCommandLine`. Something like this may work (untested):
``` nix
# The definition for the systemd service
{
  systemd.user.services.wlsunset.Service = {
     ExecStart = let
        args = cli.toGNUCommandLine {
           t = cfg.temperature.night;
           T = cfg.temperature.day;
           g = cfg.gamma;
           l = cfg.latitude;
           L = cfg.longitude;
           S = cfg.time.sunrise;
           s = cfg.time.sunset;
           d = cfg.time.duration;
           o = cfg.output;
       };
       in "${cfg.package}/bin/wlsunset ${args}";
   };
}
```

Reply:
Thank you for the recommendation. With a small tweak to it, it works perfectly
```nix
# The definition for the systemd service
{
  systemd.user.services.wlsunset.Service = {
     ExecStart = let
        args = cli.toGNUCommandLineShell {} {
           t = cfg.temperature.night;
           T = cfg.temperature.day;
           g = cfg.gamma;
           l = cfg.latitude;
           L = cfg.longitude;
           S = cfg.sunrise;
           s = cfg.sunset;
           o = cfg.output;
       };
       in "${cfg.package}/bin/wlsunset ${args}";
   };
}
```

Now it's just finding a good way to make sure that either `latitude and longitude` or `sunset and sunrise` have values.

Reply:
Now it's just finding a good way to make sure that either `latitude and longitude` or `sunset and sunrise` have values, otherwise the service won't work. Something like 
```nix
mkIf (cfg.latitude & cfg.longitude) or (cfg.sunset & cfg.sunrise) # of course this is a terrible example
```

Reply:
> Now it's just finding a good way to make sure that either `latitude and longitude` or `sunset and sunrise` have values, otherwise the service won't work. Something like
> 
> ```nix
> mkIf (cfg.latitude & cfg.longitude) or (cfg.sunset & cfg.sunrise) # for example
> ```

Something like this should work, I think: https://github.com/NixOS/nixpkgs/blob/7bb2ccd8cdc44c91edba16c48d2c8f331fb3d856/nixos/modules/services/networking/xray.nix#L72-L77

I.e. an assertion that checks that either one or the other is set, but not both.

(sorry for the late response)

Reply:
@Daru-san Will you apply the `cli.toGNUCommandLineShell` change?

Reply:
I will do that actually, I seemed to have forgotten to do that

Reply:
@matrss Looks like that works perfectly, I spent a very long time trying to find a way to that that but it seems like this simple additions does the trick
```nix
{
  assertion = (cfg.sunrise != null && cfg.sunset != null) != (cfg.latitude != null && cfg.longitude != null);
  message = "Either `sunset` and `sunrise` or `latitude` and `longitude` have to be set for wlsunset";
}

```

Reply:
I also wonder if having the options `latitude` and `longitude` and `gamma` free to be either strings, floats or integers since cli.toGNUCommandLineShell converts them all
```nix
      type = with types; nullOr (either float (either str int));
```

Reply:
Just FYI, to find that example I searched the NixOS options for the word "either" with the expectation that there must be some module with the same constraint and that word must be somewhere in the options description. So that approach can be a good starting point. Also, just grep'ing through nixpkgs can be very informative.

That said, I think your assertion still has an issue. Namely, it wouldn't catch the case that 3 of the 4 options are set, I think.

Reply:
I see, I tried experimenting with it to see how far I could go with it and it seems like there is a way, but I feel it's a bit complicated. There might be an easier way, I can try looking through nixpkgs to see what I can find
```nix
 {
   assertion = (cfg.sunrise != null && cfg.sunset != null) != (cfg.latitude != null && cfg.longitude != null)
     && (cfg.sunrise != null || cfg.sunrise != null) != (cfg.latitude != null && cfg.longitude != null);
   message = "Either `sunset` and `sunrise` or `latitude` and `longitude` have to be set for wlsunset";
 }
```

Reply:
I think you're conflating checking that `sunrise` and `sunset` have to be both `null` or set (same for `latitude`/`longitude`) with the assertion that `sunrise`/`sunset` is incompatible with `latitude`/`longitude`.

In my mind, those are three different assertions, and make the code that much easier to write.

```
 {
   assertion = (cfg.sunrise != null || cfg.sunset != null) != (cfg.latitude != null || cfg.longitude != null);
 }
 {
   assertion = (cfg.sunrise != null) == (cfg.sunset != null);
 }
 {
   assertion = (cfg.latitude != null) == (cfg.longitude != null);
 }
 ```

Reply:
That would make it way more readable, I wanted to try to make it one assertion at best but that approach is a lot easier to understand

Reply:
FWIW, I would have probably written something like
```
(cfg.sunrise != null && cfg.sunset != null && cfg.latitude == null && cfg.longitude == null) || (cfg.sunrise == null && cfg.sunset == null && cfg.latitude != null && cfg.longitude != null)
```
to make the two cases that are correct super obvious.

But splitting it into 3 assertions is even better, since you can also set meaningful messages for each of the three error cases.

Reply:
I agree, it works the best out of the other options. I will use that approach here